### PR TITLE
Update IGameEvent

### PR DESCRIPTION
**Pull request by etojuice**
Credits: 
@komashchenko for his update on igameevents.h posted in AM discord.
@hzqst GetPlayerSlot parameters

### CODE REVIEW

**Comment by GAMMACASE** [comment]
Was the ``generichash.cpp`` code copy & pasted from csgo leaked code?

File: public/igameevents.h [hunk 1]
```diff
@@ -56,6 +57,40 @@ Valid data types are string, float, long, short, byte & bool. If a
...
+		size_t len = strlen(keyName);
+
+		CBufferStringGrowable<128> hash;
+		hash.Insert(0, keyName, len);
+		hash.ToLowerFast(0);
+
+		m_nHashCode = MurmurHash2(hash.Get(), len, 0x31415926);
+		m_pszKeyName = keyName;
+
+		hash.Purge(0);
```
Comment (GAMMACASE) [inline public/igameevents.h]: The whole CBufferStringGrowable could be simplified to:
```
CUtlString buf( keyName );
buf.ToLowerFast();
m_nHashCode = MurmurHash2(buf.Get(), strlen(keyName), 0x31415926);
```

**Comment by etojuice** [comment]
> Was the `generichash.cpp` code copy & pasted from csgo leaked code?

this I can neither confirm nor deny

**Comment by GAMMACASE** [comment]
@psychonic well I'd need your approval on the code in the pr then

File: public/igameevents.h [hunk 2]
```diff
@@ -1,187 +1,225 @@
...
+{
+	inline GameEventKeySymbol_t(const char* keyName)
+	{
+		m_nHashCode = 0;
+		m_pszKeyName = keyName;
+		
+		if (!keyName || !keyName[0])
+			return;
+
+		size_t len = strlen(keyName);
```
Comment (GAMMACASE) [inline public/igameevents.h]: You don't need the len here then.

File: public/igameevents.h [hunk 3]
```diff
@@ -1,187 +1,225 @@
...
+Valid data types are string, float, long, short, byte & bool. If a 
+data field should not be broadcasted to clients, use the type "local".
+*/
+
+struct GameEventKeySymbol_t
+{
+	inline GameEventKeySymbol_t(const char* keyName)
+	{
+		m_nHashCode = 0;
+		m_pszKeyName = keyName;
```
Comment (GAMMACASE) [inline public/igameevents.h]: Also these could be moved to an initializer list instead

Comment (GAMMACASE) [reply public/igameevents.h]: You haven't moved them to initializer list, but removed, that's not really what was asked?

File: tier1/generichash.cpp [hunk 4]
```diff
@@ -301,3 +301,53 @@ unsigned FASTCALL HashBlock( const void *pKey, unsigned size )
 	return (even << 8) | odd;
 }
 
+uint32 MurmurHash2( const void *key, int len, uint32 seed )
```
Comment (psychonic) [inline tier1/generichash.cpp:304]: Pasting leaked code into our SDK isn't acceptable.

**Review by psychonic — DISMISSED** [review]

File: tier1/generichash.cpp (hunk 4 above)
Comment (xen-000) [reply tier1/generichash.cpp:304]: For posterity, this is literally in the 2013 sdk. I have no idea why @etojuice didn't bother to look there first.
https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/tier1/generichash.cpp#L313

**Comment by GAMMACASE** [comment]
@psychonic so if all is okay, this could be merged in then.

**Comment by psychonic** [comment]
Yes, although it's still marked as a draft. Were there more changes planned in this MR?

**Comment by etojuice** [comment]
> Yes, although it's still marked as a draft. Were there more changes planned in this MR?

I'm still working on this MR. I would request a review once I'm okay with the changes.

**Comment by hzqst** [comment]
GetPlayerSlot should be `virtual CPlayerSlot GetPlayerSlot(const GameEventKeySymbol_t& keySymbol) = 0;`

File: public/igameevents.h [hunk 5]
```diff
@@ -1,187 +1,218 @@
...
+
+	/* These function prototypes and names are very speculative and might be incorrect */
+	virtual CEntityHandle GetEHandle(const GameEventKeySymbol_t& keySymbol, CEntityHandle defaultValue) = 0;
+	virtual CEntityHandle GetStrictEHandle(const GameEventKeySymbol_t& keySymbol, CEntityHandle defaultValue) = 0;
+	virtual CEntityHandle GetEHandle2(const GameEventKeySymbol_t& keySymbol, CEntityHandle defaultValue) = 0;
+
+	virtual void GetPlayerSlot(CPlayerSlot &slot, const GameEventKeySymbol_t& keySymbol) = 0;
+
+	virtual void* GetPlayerController(const GameEventKeySymbol_t& keySymbol) = 0;
+	virtual void* GetPlayerPawn(const GameEventKeySymbol_t& keySymbol) = 0;
```
Comment (xen-000) [inline public/igameevents.h]: Might as well make these return `CBasePlayerController*`

Comment (GAMMACASE) [reply public/igameevents.h]: Why? We won't be having these implemented at all, just return ``IHandleEntity *`` or ``CBaseEntity *`` there and let others make a use of this?

Comment (xen-000) [reply public/igameevents.h]: I see, in that case it makes more sense to return what you said.

**Comment by GAMMACASE** [comment]
@etojuice I've updated the interface with my findings as well, now should be more complete, can you revisit it? And if all is okay I'll merge that in.

**Comment by etojuice** [comment]
> @etojuice I've updated the interface with my findings as well, now should be more complete, can you revisit it? And if all is okay I'll merge that in.

lgtm. ty

**Pull request merged**